### ISSUE TITLE
htmlFor attribute on labels not working with VoiceOver in Safari

### ISSUE THREAD

**Issue by francinen** [facebook/react#12905]
**Do you want to request a *feature* or report a *bug*?**
A bug

**What is the current behavior?**
The `htmlFor` attribute on `label`s is not behaving consistently across browsers when VoiceOver is enabled.

In Chrome, the label value is announced by VoiceOver when an input receives focus (as expected): https://www.youtube.com/watch?v=tAFX3SWFYbY&feature=youtu.be

In Safari, however, the label value is not announced by VoiceOver: https://www.youtube.com/watch?v=cQ-0HuMf0bk&feature=youtu.be

I was able to get VoiceOver to announce the expected label in Safari by adding an `aria-labelledby` attribute to the inputs:
```javascript
<label htmlFor="FullName" id="FullNameLabel">Full name</label>
<input type="text" id="FullName" aria-labelledby="FullNameLabel" onChange={this.onNameChange} value={this.state.name} />
<label htmlFor="EmailAddress" id="EmailAddressLabel">Email address</label>
<input type="text" id="EmailAddress" aria-labelledby="EmailAddressLabel" onChange={this.onEmailChange} value={this.state.email} />
<label htmlFor="Info" id="InfoLabel">Are there any considerations you would like to share with the instructor (e.g., injuries, sensitivities, requests)?</label>
<textarea id="Info" aria-labelledby="InfoLabel" onChange={this.onInfoChange} value={this.state.info} rows="5" />
```

**What is the expected behavior?**
Using `htmlFor` should suffice to get VoiceOver to read the expected label value across all browsers.

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**
React 16.3.2
Chrome and Safari on MacOS High Sierra with VoiceOver enabled

**Comment by svinkle** [comment]
This is actually a bug with VO + Safari + `role="dialog"` + `aria-modal="true"`.

Any `label` + `input` pairing just doesn't get announced within the dialog context. Can't recall if this was ever filled… 🤔

See [Demo on CodePen](https://codepen.io/svinkle/pen/GdbPNq).

**Comment by gaearon** [comment]
Sounds like not a React issue then.
Let me know if I’m missing something!
